plugins/builtin/target/aws-asg/plugin/aws.go

```diff
@@ -93,105 +93,168 @@ func (t *TargetPlugin) scaleOut(ctx context.Context, asg *autoscaling.AutoScalin
+		return err
+	}
+	t.logger.Info("pre scale-in tasks now complete")
+
 	// Create the event writer and write that the drain event has been
 	// completed which is part of the RunPreScaleInTasks() function.
 	eWriter := newEventWriter(t.logger, t.asg, instanceIDs, *asg.AutoScalingGroupName)
 	eWriter.write(ctx, scalingEventDrain)
 
 	// Create a logger for this action to pre-populate useful information we
 	// would like on all log lines.
-	log := t.logger.With("action", "scale_in", "asg_name", *asg.AutoScalingGroupName,
-		"instances", instanceIDs)
+	log := t.logger.With("action", "scale_in", "asg_name", *asg.AutoScalingGroupName)
 
-	// Detach the desired instances.
-	log.Debug("detaching instances from AutoScaling Group")
+	// Run the termination and log the results.
+	result := t.terminateInstancesInASG(ctx, nodeResourceIDs)
+	result.logResults(log)
 
-	if err := t.detachInstances(ctx, asg.AutoScalingGroupName, instanceIDs); err != nil {
-		return fmt.Errorf("failed to scale in AWS AutoScaling Group: %v", err)
+	// If we have any failures, perform our revert so we don't leave nodes in
+	// an undesired state.
+	if result.lenFailure() > 0 {
+		t.clusterUtils.RunPostScaleInTasksOnFailure(result.failedIDs())
 	}
-	log.Info("successfully detached instances from AutoScaling Group")
-	eWriter.write(ctx, scalingEventDetach)
 
-	// Terminate the detached instances.
-	log.Debug("terminating EC2 instances")
+	// If we had successfully termination from the ASG, track these activities
+	// until completion. A failure here should not fail the scaling activity as
+	// AWS should honour the contract, it could be a case of there being
+	// slowness in the AWS system and us timing out.
+	if result.lenSuccess() > 0 {
 
-	if err := t.terminateInstances(ctx, instanceIDs); err != nil {
-		return fmt.Errorf("failed to scale in AWS AutoScaling Group: %v", err)
-	}
-	log.Info("successfully terminated EC2 instances")
-	eWriter.write(ctx, scalingEventTerminate)
+		t.logger.Debug("ensuring AWS ASG activities complete")
+
+		if err := t.ensureActivitiesComplete(ctx, *asg.AutoScalingGroupName, result.activityIDs()); err != nil {
+			log.Error("failed to ensure all activities completed", "error", err)
+		} else {
+			t.logger.Debug("confirmed AWS ASG activities completed")
+		}
+		eWriter.write(ctx, scalingEventTerminate)
 
-	// Run any post scale in tasks that are desired.
-	if err := t.clusterUtils.RunPostScaleInTasks(ctx, config, ids); err != nil {
-		return fmt.Errorf("failed to perform post-scale Nomad scale in tasks: %v", err)
+		// Run any post scale in tasks that are desired.
+		if err := t.clusterUtils.RunPostScaleInTasks(ctx, config, result.successfulIDs()); err != nil {
+			return fmt.Errorf("failed to perform post-scale Nomad scale in tasks: %v", err)
+		}
 	}
 
-	return nil
+	if result.lenFailure() > 0 && result.lenSuccess() > 0 {
+		log.Warn("partial scaling success",
+			"success_num", result.lenSuccess(), "failed_num", result.lenFailure())
+		return nil
```

Comment:
I would appreciate peoples view on this part in particular!

Reply:
so, as long as we had at least one successful scale in, we don't return an error here, with the assumption that we'll retry.

one thing that i'm not sure about is that we don't keep the error if we failed to reconcile the node state; `RunPostScaleInTasksOnFailure` doesn't return an error. so it's possible that we got to this point and we have nodes that were not destroyed but are still ineligible. any thoughts on that? should that error case prevent us from returning a null on partial success?

Reply:
Ah yes this is a great idea.

So collect any errors from `RunPostScaleInTasksOnFailure`. In the event there are some and we have a partial result, return these errors to the caller specifying the scaling action failed because we were unable to reconcile the state thus indicating the operator needs to take action.